Strip unicode control characters from TEXT before entering the database

mvdbeek opened the pull request:
I'm seeing this on my dev and prod instance:
```
galaxy.jobs.runners ERROR 2019-05-03 14:01:43,578 (953/) Job wrapper finish method failed
Traceback (most recent call last):
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/runners/__init__.py", line 509, in _finish_or_resubmit_job
    job_wrapper.finish(tool_stdout, tool_stderr, exit_code, check_output_detected_state=check_output_detected_state, job_stdout=job_stdout, job_stderr=job_stderr)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/jobs/__init__.py", line 1385, in finish
    self.external_output_metadata.load_metadata(dataset, dataset_assoc.name, self.sa_session, working_directory=self.working_directory, remote_metadata_directory=remote_metadata_directory)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/metadata/__init__.py", line 153, in load_metadata
    self._load_metadata_from_path(dataset, metadata_output_path, working_directory, remote_metadata_directory)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/metadata/__init__.py", line 79, in _load_metadata_from_path
    dataset.metadata.from_JSON_dict(metadata_output_path, path_rewriter=path_rewriter)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/model/metadata.py", line 182, in from_JSON_dict
    value = param.from_external_value(external_value, dataset, **from_ext_kwds)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/model/metadata.py", line 544, in from_external_value
    mf = self.new_file(dataset=parent, **value.kwds)
  File "/Users/mvandenb/src/galaxy/lib/galaxy/model/metadata.py", line 579, in new_file
    sa_session.flush()  # flush to assign id
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2446, in flush
    self._flush(objects)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2584, in _flush
    transaction.rollback(_capture_exception=True)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/langhelpers.py", line 67, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 277, in reraise
    raise value
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/session.py", line 2544, in _flush
    flush_context.execute()
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 416, in execute
    rec.execute(self)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/unitofwork.py", line 583, in execute
    uow,
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 236, in save_obj
    update,
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/orm/persistence.py", line 976, in _emit_update_statements
    statement, multiparams
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 980, in execute
    return meth(self, multiparams, params)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/sql/elements.py", line 273, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1099, in _execute_clauseelement
    distilled_params,
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1240, in _execute_context
    e, statement, parameters, cursor, context
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1460, in _handle_dbapi_exception
    util.reraise(*exc_info)
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/util/compat.py", line 277, in reraise
    raise value
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/base.py", line 1236, in _execute_context
    cursor, statement, parameters, context
  File "/Users/mvandenb/src/galaxy/.venv3/lib/python3.6/site-packages/sqlalchemy/engine/default.py", line 536, in do_execute
    cursor.execute(statement, parameters)
ValueError: A string literal cannot contain NUL (0x00) characters.
```

The docker based tests have no issue, and this also seems to work fine on docker-galaxy-stable.
So I'm not quite sure if we want to include this, or if there is some config setting that I'm missing.

Fixes https://github.com/bgruening/docker-galaxy-stable/issues/489 and https://github.com/galaxyproject/tools-iuc/issues/2438

mvdbeek committed:
Change unicode test tool

To produce NULL character. This tool fails locally but not
if testing against the jenkins framework tests or using
external docker-galaxy-stable. I don't understand why.

mvdbeek commented:
That said
```
hda = HistoryDatasetAssociation()
hda.tool_version = '\0x00'
sa_session.add(hda)
sa_session.flush()
```
blows up everywhere, so this is probably the sane thing to do, unless we migrate to a column type that doesn't have this issue.

abretaud commented:
I have the same error with Bowtie2 (xref https://github.com/bgruening/docker-galaxy-stable/issues/489 and https://github.com/galaxyproject/tools-iuc/issues/2438) after updating my dev instance to 19.05 (I didn't test with 19.01)
I don't really understand which strange character triggers this error... Is there a place I should look at? I have a job directory at hand if needed (file in metadata dir?)

mvdbeek committed:
Strip unicode null from stdout/stderr/tool_version

mvdbeek committed:
Strip unicode null from peek

mvdbeek commented:
After discussing this with @nsoranzo on gitter a while ago we've decided it would be best to to strip null characters when we encounter them first, so I've changed `unicodify` to optionally strip `\0` (the offending character here). We do this for `tool_stdout`, `tool_stderr`, `job_stdout`, `job_stderr`. For `peek` that is may be set in many different datatypes I've moved the column-mapped attribute to`_peek` and used a getter/setter to sanitize the `peek` value.
I think this should cover all cases where we can reasonably expect a `TEXT` column to contain `\0`.

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -289,11 +289,13 @@ def metrics(self):
         def shrink_and_unicodify(what, stream):
-            stream = galaxy.util.unicodify(stream) or u''
-            if (len(stream) > galaxy.util.DATABASE_MAX_STRING_SIZE):
-                stream = galaxy.util.shrink_string_by_size(tool_stdout, galaxy.util.DATABASE_MAX_STRING_SIZE, join_by="\n..\n", left_larger=True, beginning_on_size_error=True)
```
Also this was a bug imo, it should have been `stream`, not `tool_stdout`. So for large stderr streams we'd get the stdout again ...

nsoranzo commented on lib/galaxy/util/__init__.py:
```diff
@@ -1008,7 +1019,11 @@ def unicodify(value, encoding=DEFAULT_ENCODING, error='replace'):
+    if isinstance(value, text_type):
+        if strip_null:
+            return value.replace('\0', '')
         return value
```
If we want to simplify the code, we can remove the shortcut for `text_type` (i.e. lines 1024-1026).

mvdbeek replied:
Good point, let's do that, I doubt this is performance critical.

mvdbeek committed:
Simplify unicodify

Thanks Nicola!
